### PR TITLE
Add `npm run quality` script, Update Readme

### PR DESCRIPTION
**Pull request by mukkachaitanya**
### Requirements

* Need for easy and better checks for code quality. 

### Description of the Change

Adds `npm run quality` script that uses eslint, jsinspect and buddy.js to keep a check on the quality of the project. 

### Benefits

Improved and easy code quality checking.

### Applicable Issues

Addresses #96

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -9,7 +9,8 @@
     "integration": "nyc --reporter=lcov --reporter=text-lcov -s --all mocha 'test/integration/**/*.js'",
     "post-integration": "nyc report --reporter=lcov > coverage.lcov && codecov -F integration && rm -rf ~/.autolabjs",
     "feature": "nyc --reporter=lcov --reporter=text-lcov -s --all cucumber-js 'test/feature/features' -r 'test/feature/steps'",
-    "post-feature": "nyc report --reporter=lcov > coverage.lcov && codecov -F feature"
+    "post-feature": "nyc report --reporter=lcov > coverage.lcov && codecov -F feature",
+    "quality" : "eslint . & jsinspect . & buddy ."
```
Comment (prasadtalasila) [inline package.json]: now the style is to put a comma even on the last line so that the diff shows modification only on one line.

Comment (prasadtalasila) [reply package.json]: "quality" : "eslint . && jsinspect . && buddy ."

File: .eslintrc.json [hunk 2]
```diff
@@ -1,5 +1,9 @@
 {
     "extends": "airbnb-base",
+    "env" : {
```
Comment (prasadtalasila) [inline .eslintrc.json:3]: why are these needed?

File: .jsinspectrc [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "threshold":     60,
```
Comment (prasadtalasila) [inline .jsinspectrc]: This is not similar to codeclimate duplication weight. This number is the number of lines. A prudent value seems to be between 4 to 6.

**Review by prasadtalasila — CHANGES_REQUESTED** [review]
@mukkachaitanya please see the suggested changes.

**Comment by prasadtalasila** [comment]
@mukkachaitanya also it is better to run `npm quality` command in the travis after the tests run successfully.

File: .eslintrc.json (hunk 2 above)
Comment (mukkachaitanya) [reply .eslintrc.json:3]: `eslint` give errors of the form : 
```
12:3   error  'it' is not defined                                            no-undef
18:3   error  'describe' is not defined                                      no-undef
```

when `mocha` env is not set to be true.

File: .jsinspectrc (hunk 3 above)
Comment (mukkachaitanya) [reply .jsinspectrc]: I'll make that change.

**Comment by mukkachaitanya** [comment]
Also Sir, there are a lot of `eslint` errors which have to be taken care of. I don't see codeclimate reporting them.

File: package.json (hunk 1 above)
Comment (mukkachaitanya) [reply package.json]: Sir, if `eslint` fails the none of the further checks are taken, i.e., none of jsinspect/buddy commands are executed.

Comment (mukkachaitanya) [reply package.json]: Also sir, `json` doesn't accept a trailing comma. Throws the following error :
```
npm ERR! JSON.parse Unexpected token } in JSON at position 831 while parsing near '...ect . & buddy .",
npm ERR! JSON.parse   },
npm ERR! JSON.parse   "bin": {
npm ERR! JSON.parse     "a...'
npm ERR! JSON.parse Failed to parse package.json data.
npm ERR! JSON.parse package.json must be actual JSON, not just JavaScript.
```

**Comment by prasadtalasila** [comment]
> Also Sir, there are a lot of `eslint` errors which have to be taken care of. I don't see codeclimate reporting them.

This is rude awakening for me as well. Please fix these issues before we do further development. One hypersensitive check is the threshold parameter in `.jsinspectrc`. This number can be increased to 20.

**Comment by mukkachaitanya** [comment]
Some changes in commit 6db57af to be considered :

- Added `buddyrc` file to configure the `buddy.js` tool, which ignores `[0, 1, 2, 3]` as constants as we use them in the array indexing.
- The threshold for `jsinspect` has been changed to the default value `30` as it was reporting issuses which we discussed to not to be limit.
- The following rule was [modified ](https://github.com/AutolabJS/autolabcli/blob/6db57afda71a24c4a187d218135de3d176fcabc7/.eslintrc#L9)to allow the usage of unary `++` or `--` operators in loops only.
```js
"no-plusplus" : ["error", { "allowForLoopAfterthoughts": true }],
``` 

-  Also, the following rule was [added ](https://github.com/AutolabJS/autolabcli/blob/6db57afda71a24c4a187d218135de3d176fcabc7/.eslintrc#L10)to take care of issues arising from `cumumber.js` tests
```js
"no-unused-vars": [
            "error",
            {
              "varsIgnorePattern": "should|expect|Before|After|Given"
            }
        ]
```
- The following rule also switches of `max-lines-per-function` in the [unit tests](https://github.com/AutolabJS/autolabcli/blob/6db57afda71a24c4a187d218135de3d176fcabc7/test/.eslintrc)
```js
 "max-lines-per-function" : ["off"]
```
- Also the the rule `no-reassign-param` has been switched off in some places especially in the integration test. (for example, [test/integration/eval.js](https://github.com/AutolabJS/autolabcli/blob/6db57afda71a24c4a187d218135de3d176fcabc7/test/integration/eval.js#L42))

File: .buddyrc [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "detectObjects": false,
+  "enforceConst":  false,
+  "ignore":        [0, 1, 2, 3],
```
Comment (prasadtalasila) [inline .buddyrc]: please remove this ignore statement.

File: .buddyrc [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
+  "detectObjects": false,
```
Comment (prasadtalasila) [inline .buddyrc]: please clarify the use of next two lines.

File: .eslintrc [hunk 6]
```diff
@@ -0,0 +1,17 @@
+{
+    "extends": "airbnb-base",
+    "env" : {
+        "node": true,
+        "mocha": true
+    },
+    "rules": {
```
Comment (prasadtalasila) [inline .eslintrc]: please clarify the reason for using the rule.

File: .eslintrc [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+        "max-lines-per-function" : ["error", 30],
+        "no-plusplus" : ["error", { "allowForLoopAfterthoughts": true }],
+        "no-unused-vars": [
+            "error",
+            {
+              "varsIgnorePattern": "should|expect|Before|After|Given"
```
Comment (prasadtalasila) [inline .eslintrc]: avoid this rule using a plug-in?

File: lib/cli/output/eval.js [hunk 8]
```diff
@@ -37,9 +41,9 @@ const printTableAndGetScore = (data) => {
 };
 
 const printStatusAndLog = (data, totalScore) => {
-  logger.moduleLog('debug', 'Eval Output', `Evaluation logs: ${new Buffer(data.log, 'base64').toString()}.`);
+  logger.moduleLog('debug', 'Eval Output', `Evaluation logs: ${Buffer.from(data.log, 'base64').toString()}.`);
```
Comment (prasadtalasila) [inline lib/cli/output/eval.js:44]: ?

File: lib/cli/input/prefs.js [hunk 9]
```diff
@@ -87,16 +87,20 @@ const changeGitlab = (host) => {
 };
 
 const changeServer = async (options) => {
-  let { host, port, type } = options;
+  const { host, port } = options;
+  let { type } = options;
   if (!type) {
     const typePrompt = prefPrompts.getServerTypePrompt();
-    type = (await inquirer.prompt(typePrompt)).type;
+    ({ type } = (await inquirer.prompt(typePrompt)));
```
Comment (prasadtalasila) [inline lib/cli/input/prefs.js:94]: why brackets?

File: lib/model/prefs.js [hunk 10]
```diff
@@ -77,7 +80,7 @@ const storePrefs = (changePrefEvent) => {
   } else if (changePrefEvent.name === 'server_changed') {
     storeServer(changePrefEvent.details);
   } else if (changePrefEvent.name === 'show_prefs') {
-    changePrefEvent = eventForShowPref(changePrefEvent);
+    return eventForShowPref(changePrefEvent);
```
Comment (prasadtalasila) [inline lib/model/prefs.js:83]: ?

File: package.json [hunk 11]
```diff
@@ -9,7 +9,8 @@
     "integration": "nyc --reporter=lcov --reporter=text-lcov -s --all mocha 'test/integration/**/*.js'",
     "post-integration": "nyc report --reporter=lcov > coverage.lcov && codecov -F integration && rm -rf ~/.autolabjs",
     "feature": "nyc --reporter=lcov --reporter=text-lcov -s --all cucumber-js 'test/feature/features' -r 'test/feature/steps'",
-    "post-feature": "nyc report --reporter=lcov > coverage.lcov && codecov -F feature"
+    "post-feature": "nyc report --reporter=lcov > coverage.lcov && codecov -F feature",
+    "quality": "eslint . & jsinspect . & buddy ."
```
Comment (prasadtalasila) [inline package.json]: please add &&

**Review by prasadtalasila — CHANGES_REQUESTED** [review]

File: lib/utils/preference-manager.js [hunk 12]
```diff
@@ -1,3 +1,4 @@
+/* eslint no-use-before-define: "warn" */
```
Comment (mukkachaitanya) [inline lib/utils/preference-manager.js]: Need to refactor the cyclic functions.

File: .buddyrc (hunk 5 above)
Comment (mukkachaitanya) [reply .buddyrc]: According to the documentation : 
```
detectObjects    -    detect object expressions and properties
enforceConst     -    require literals to be defined using const
```

I think we can switch both to `true`.

File: lib/cli/output/eval.js (hunk 8 above)
Comment (mukkachaitanya) [reply lib/cli/output/eval.js:44]: The Buffer constructor is different depending on the type of its argument, and not validating such arguments might lead to memory disclosure and denial of service. So eslint enforces the use of `Buffer.from()` or `Buffer.alloc()` instead. 

[ref](https://eslint.org/docs/rules/no-buffer-constructor)

File: lib/cli/input/prefs.js (hunk 9 above)
Comment (mukkachaitanya) [reply lib/cli/input/prefs.js:94]: We have discussed this, sir, but for having everything in a single thread; the MDN documentation says: 

> 

    The parentheses ( ... ) around the assignment statement are required when using object literal destructuring assignment without a declaration.

    {a, b} = {a: 1, b: 2} is not valid stand-alone syntax, as the {a, b} on the left-hand side is considered a block and not an object literal.

    However, ({a, b} = {a: 1, b: 2}) is valid, as is var {a, b} = {a: 1, b: 2}

    NOTE: Your ( ... ) expression needs to be preceded by a semicolon or it may be used to execute a function on the previous line.

File: lib/model/prefs.js (hunk 10 above)
Comment (mukkachaitanya) [reply lib/model/prefs.js:83]: The function `storePrefs` was modifying the input argument `changePrefEvent` storing and then returning it. Instead, the above modification returns the new value directly.
